File: Data Structures & Algorithms/search-2d-matrix/submission-0.js
class Solution {
    /**
     * @param {number[][]} matrix
     * @param {number} target
     * @return {boolean}
     */
    searchMatrix(matrix, target) {

        const numRows = matrix.length;
        const numCols = matrix[0].length;

        let left = 0, right = numRows * numCols - 1;

        while(left <= right)
        {
            const mid = left + Math.floor((right - left)/2);

            const row = Math.floor(mid / numCols);
            const col = mid % numCols;

            if(matrix[row][col] == target) return true;
            else if(matrix[row][col] < target) left = mid + 1;
            else right = mid - 1;
        }

        return false;
    }
}

File: Data Structures & Algorithms/set-zeroes-in-matrix/submission-0.cpp
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        
        int numRows = matrix.size(), numCols = matrix[0].size(), rfr = 0, rfc = 0;

        for(int i=0;i<numRows;i++)
        {
            for(int j=0;j<numCols;j++)
            {
                if(matrix[i][j] == 0)
                {
                    if(i == 0) rfr = 1;
                    if(j == 0) rfc = 1;

                    matrix[0][j] = 0;
                    matrix[i][0] = 0;
                }
            }
        }

        for(int i=1;i<numRows;i++)
        {
            for(int j=1;j<numCols;j++)
            {
                if(matrix[i][0] == 0 || matrix[0][j] == 0) matrix[i][j] = 0;
            }
        }

        if(rfr) {
            for(int i=0;i<numCols;i++) matrix[0][i] = 0;
        }

        if(rfc) {
            for(int i=0;i<numRows;i++) matrix[i][0] = 0;
        }
    }
};
